framework/src/mesh/MooseMesh.C

```diff
@@ -1237,61 +1238,99 @@ MooseMesh::detectOrthogonalDimRanges(Real tol)
-   */
-  unsigned int strides[3][3] = {{4, 2, 1}, {2, 4, 1}, {1, 4, 2}};
-
-  // Grab a reference to the BoundaryInfo object
-  BoundaryInfo & boundary_info = getMesh().get_boundary_info();
-
-  buildNodeListFromSideList();
-
+  // Loop over level-0 elements (since boundary condition information
+  // is only directly stored for them) and find sidesets with normals
+  // that point in the -x, +x, -y, +y, and -z, +z direction.  If there
+  // is a unique sideset id for each direction, then the paired
+  // sidesets consist of (-x,+x), (-y,+y), (-z,+z).  If there are
+  // multiple sideset ids for a given direction, then we can't pick a
+  // single pair for that direction.  In that case, we'll just return
+  // as was done in the original algorithm.
+  std::set<BoundaryID>
+    minus_x_ids, plus_x_ids,
+    minus_y_ids, plus_y_ids,
+    minus_z_ids, plus_z_ids;
+
+  // Points used for direction comparison
+  const Point
+    minus_x(-1,  0,  0),
+    plus_x ( 1,  0,  0),
+    minus_y( 0, -1,  0),
+    plus_y ( 0,  1,  0),
+    minus_z( 0,  0, -1),
+    plus_z ( 0,  0,  1);
+
+  // A first-order Lagrange FE for the face.
   unsigned int dim = getMesh().mesh_dimension();
-  unsigned int z_dim = dim == 3 ? 2 : 1;  // Determine how many z loops there are, we always need to loop at least once!
+  UniquePtr<FEBase> fe_face (FEBase::build(dim,
+                                           FEType(FIRST, LAGRANGE)));
 
-  // Container to catch IDs passed back from the BoundaryInfo object
-  std::vector<BoundaryID> bounds_ids;
+  // Face is assumed to be flat, therefore normal is assumed to be
+  // constant over the face, therefore only compute it at 1 qp.
+  QGauss qface(dim-1, CONSTANT);
+  fe_face->attach_quadrature_rule (&qface);
+  const std::vector<Point> & normals = fe_face->get_normals();
 
-  for (unsigned int curr_dim = 0; curr_dim < dim; ++curr_dim)
+  // We need this to get boundary ids for each boundary face we encounter.
+  BoundaryInfo & boundary_info = getMesh().get_boundary_info();
+  std::vector<boundary_id_type> face_ids;
+
+  MeshBase::const_element_iterator       el     = getMesh().level_elements_begin(0);
+  const MeshBase::const_element_iterator end_el = getMesh().level_elements_end(0);
+  for (; el != end_el ; ++el)
   {
-    std::pair<BoundaryID, BoundaryID> paired_boundary;
-    for (unsigned int i = 0; i < 2; ++i)
-    {
-      std::map<BoundaryID, unsigned int> boundary_counts;
-      for (unsigned int j = 0; j < 2; ++j)
-        for (unsigned int k = 0; k < z_dim; ++k)
-        {
-          boundary_info.boundary_ids(_extreme_nodes[i*strides[curr_dim][0] + j*strides[curr_dim][1] + k*strides[curr_dim][2]], bounds_ids);
+    Elem * elem = *el;
 
-          for (unsigned int l = 0; l<bounds_ids.size(); ++l)
-            ++boundary_counts[bounds_ids[l]];
-        }
+    // loop over element sides
+    for (unsigned int s=0; s<elem->n_sides(); s++)
```

Comment:
Oh yeah, I copy/pasted this code from somewhere else.  I will do a fixup commit.